testing/unique.cu

```diff
@@ -266,3 +310,53 @@ struct TestUniqueCopyToDiscardIterator
 VariableUnitTest<TestUniqueCopyToDiscardIterator, IntegralTypes> TestUniqueCopyToDiscardIteratorInstance;
 
 
+template <typename Vector>
+void TestUniqueCountSimple(void)
+{
+    typedef typename Vector::value_type T;
+
+    Vector data(10);
+    data[0] = 11;
+    data[1] = 11;
+    data[2] = 12;
+    data[3] = 20;
+    data[4] = 29;
+    data[5] = 21;
+    data[6] = 21;
+    data[7] = 31;
+    data[8] = 31;
+    data[9] = 37;
+
+    int count = thrust::unique_count(data.begin(), data.end());
+
+    ASSERT_EQUAL(count, 7);
+
+    int div_10_count = thrust::unique_count(data.begin(), data.end(), is_equal_div_10_unique<T>());
+
+    ASSERT_EQUAL(div_10_count, 3);
+}
+DECLARE_INTEGRAL_VECTOR_UNITTEST(TestUniqueCountSimple);
```

Comment:
`unique_count` claims to work with forward iterators. There should be a test using forward iterators.

Reply:
Is there a nice way to wrap an iterator into a forward_iterator in Thrust? I wrote a small wrapper class and that seems to work and compile, but I suppose that problem has been solved elsewhere already?

Reply:
@ericniebler is right, we should be testing this. Unfortunately we lack robust testing for these sorts of things.

Thanks for adding the new testing infrastructure! Please include it in this PR, ideally in the testing framework so we can reuse it from other tests later 👍

Reply:
This should be possible. Thrust packs some extra info into its iterator category. See `thrust/iterator/iterator_categories.h` for more info.

The iterator wrapper will need to test the base iterator using `thrust::iterator_system<BaseIterator>::type`, and define its iterator category to `forward_host_iterator_tag` if the result is `host_system_tag`, or `forward_device_iterator_tag` if the result is `device_iterator_tag`.

Reply:
This should be possible. Thrust packs some extra info into its iterator category. See `thrust/iterator/iterator_categories.h` for more info.

The iterator wrapper will need to test the base iterator using `thrust::iterator_system<BaseIterator>::type`, and define its iterator category to `forward_host_iterator_tag` if the result is `host_system_tag`, or `forward_device_iterator_tag` if the result is `device_iterator_tag`.

Reply:
I tried using this approach, but it looks like the implementation internally still tries to use the forward_iterator_wrapper like a random access iterator, so the compilation fails missing operator[] and operator+ and similar places.

Corresponding diff:
```diff
diff --git a/testing/unittest/iterator_helpers.h b/testing/unittest/iterator_helpers.h
index 8d4f03f5..8f254f46 100644
--- a/testing/unittest/iterator_helpers.h
+++ b/testing/unittest/iterator_helpers.h
@@ -1,6 +1,8 @@
 #pragma once
 
 #include <thrust/iterator/iterator_traits.h>
+#include <thrust/iterator/iterator_categories.h>
+#include <type_traits>
 
 
 // Wraps an existing iterator into a forward iterator,
@@ -8,11 +10,18 @@
 template <typename Iterator>
 struct forward_iterator_wrapper {
     // LegacyIterator requirements
+    using iterator_system_tag = typename thrust::iterator_system<Iterator>::type;
     using reference = typename thrust::iterator_traits<Iterator>::reference;
     using pointer = typename thrust::iterator_traits<Iterator>::pointer;
     using value_type = typename thrust::iterator_traits<Iterator>::value_type;
     using difference_type = typename thrust::iterator_traits<Iterator>::difference_type;
-    using iterator_category = std::forward_iterator_tag;
+    using iterator_category = typename std::conditional<
+        std::is_convertible<iterator_system_tag, thrust::device_system_tag>::value,
+        thrust::forward_device_iterator_tag,
+        typename std::conditional<
+            std::is_convertible<iterator_system_tag, thrust::host_system_tag>::value,
+            thrust::forward_host_iterator_tag,
+            std::forward_iterator_tag>::type>;
     using base_iterator_category = typename thrust::iterator_traits<Iterator>::iterator_category;
     static_assert(
         std::is_convertible<base_iterator_category, std::forward_iterator_tag>::value,
```

Reply:
@upsj What's the current status of the dispatch issues? You'd mentioned that this was doing odd things like dispatching host_vectors to device and vice-versa -- was that resolved?

Reply:
Let me take another look at it, I haven't touched this in a while

Reply:
@allisonvacanti forgot a ::type in my std::conditional, it seems to work now :)

Reply:
This sounded odd to me, as I've never seen any logic in Thrust that would dispatch forward iterators to serial implementations. So I dug into it, and unfortunately this is due to a pretty nasty bug in Thrust's iterator traits.

The details are gory, but I've summarized them [in a comment](https://github.com/NVIDIA/thrust/issues/902#issuecomment-1119107644) on #902.

Until that's fixed, I'm not comfortable merging this test using the forward_iterator_wrapper, since they only "do the right thing" because the iterator framework is broken.

I hate to churn on this PR even more, but I think we should remove the iterator wrappers for now and just test that the regular iterators work. We can re-introduce the wrapper tests as part of #55, after #902 is fixed and settled.

@ericniebler Can you review the two linked issues and see if you agree with my suggestion?

Reply:
I don't think that forward iterators actually need to be dispatched to the sequential backend. They support multipass reading and should be usable in a parallel algorithm, so long as they're only copied and incremented. Is there something in the unique_count/count_if algorithms that would break them?

Reply:
Oof, that sounds annoying. The main issue I see with parallel execution on forward iterators is that they introduce an essentially linear dependency chain that means either every thread ´i` starts from `begin` and increments it `n` times, or waits until one of its predecessors `j` is done and writes its iterator somewhere, to then increments it `i - j` times. Both don't really seem useful for parallel execution to me.

Reply:
It will require more increments, but if the work-per-element is expensive compared to the cost of the iterator increment, it may still make sense to parallelize. I'd rather let the user make that call, since they can opt-in to sequential execution by passing in the sequential exeuction policy (`thrust::seq`).

More importantly, the sequential implementation executes on CPU, and some types of device memory aren't accessible from the CPU's address space, so switching to `seq` really needs to be opt-in rather than default.

Reply:
Thanks for the clarification, that makes sense! I was only thinking of simple but massively parallel cases.